### PR TITLE
Lay the foundation for supporting UPM packaging of MRTK

### PR DESCRIPTION
**Pull request by davidkline-ms**
This change updates the MixedRealityToolkitFiles class to support MRTK residing either the Assets (direct import) or Packages (UPM import) folders. Tests have been added for the new file path processing functionality.

One additional change is to shorten the length of the Wndows Mixed Reality Configuration Checker assembly name as it causes issues due to an increased path length when imported via UPM.

NOTE: This change does not create UPM packages. That change will follow at a later time.

### CODE REVIEW

File: Assets/MRTK/Core/AssemblyInfo.cs [hunk 1]
```diff
@@ -8,3 +8,5 @@
 
 [assembly: AssemblyProduct("Microsoft® Mixed Reality Toolkit")]
 [assembly: AssemblyCopyright("Copyright © Microsoft Corporation")]
+
+[assembly: UnityEngine.Scripting.Preserve]
```
Comment (keveleigh) [inline Assets/MRTK/Core/AssemblyInfo.cs:12]: Not sure how I feel about marking so many of our assemblies with this tag. Our existing link.xml might be a contributor to reported increased build times. https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5273#issuecomment-513974004

Comment (keveleigh) [reply Assets/MRTK/Core/AssemblyInfo.cs:12]: Can we just mark types that are used via `SystemType` serialization? It feels like that's the problem case.

Comment (davidkline-ms) [reply Assets/MRTK/Core/AssemblyInfo.cs:12]: I can do that. It will likely be the services, providers and extensions (which is the majority of those changed here).

Comment (davidkline-ms) [reply Assets/MRTK/Core/AssemblyInfo.cs:12]: That said, a pruning of the list _is_ warranted here

Comment (davidkline-ms) [reply Assets/MRTK/Core/AssemblyInfo.cs:12]: @keveleigh, do you think it is worth marking the _classes_ vs marking the entire assemblies?

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WMR.ConfigCheck.asmdef [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-    "name": "Microsoft.MixedReality.Toolkit.Providers.WindowsMixedReality.ConfigCheck",
+    "name": "Microsoft.MixedReality.Toolkit.Providers.WMR.ConfigCheck",
```
Comment (wiwei) [inline Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WMR.ConfigCheck.asmdef:2]: Do we expect other people to take dependencies on this config check assembly? If so this would be a breaking change we'd want to add to the releasenotes.md.

Why was this changed in this PR?

**Comment by wiwei** [comment]
I wanted to +1 to keveleigh@'s comment around not doing assembly-global preserve type. The reason we had to add link.xml was to solve an issue around code stripping in precompiled binaries (which obviously also happens again when dealing with UPM.

I don't think that we should be marking everything globally as preserve, since this will slow down every single distribution mechanism (including people who work on a day-to-day basis on the MRTK itself).

I think that even a class-based approach here is prone to issues, because we will likely need to make it very expansive to cover everything as you mention (i.e. anything that uses SystemType serialization). I think that unless we build more tooling (i.e. something that validates that SystemType classes are covered by the set of preserve attributes), we run the risk of really hard to diagnose bugs (we've seen the messages before, they aren't very clear), and it would be really easy to mess this up (i.e. it's super easy to forget a preserve attribute on a new class).

Note that it's possible to have a link.xml that references OTHER assemblies:

https://docs.unity3d.com/Manual/ManagedCodeStripping.html

_Note that link.xml files are not supported inside packages, but you can reference package assemblies from non-package link.xml files._

If the intent here is to unblock further development then actually modifying the link.xml to do package reference assemblies might be a semi decent near term thing (i.e. because there's more UPM work to be done that we might want to get to first).

I think that in the end getting to some class-based marking would be good (I'm assuming it only affects things in packages so we wouldn't have to worry about byte stripping for custom classes and other things people use SystemType for in their own projects) but we'd need to ensure that we have tooling to ensure that we have proper coverage going forward.

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WMR.ConfigCheck.asmdef (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WMR.ConfigCheck.asmdef:2]: We do not expect any customers to depend upon this assembly. It is used internally by MRTK.

It was changed here due to the path length getting too long and breaking compile of the Windows Mixed Reality provider

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs [hunk 3]
```diff
@@ -215,8 +215,12 @@ private static void Init()
...
         public static void RefreshFolders()
         {
-            string path = Application.dataPath;
-            searchForFoldersTask = Task.Run(() => SearchForFoldersAsync(path));
+            // MRTK may be located in Assets (.unitypackage import) or the Packages (UPM import)
+            // folder. Check both locations.
+            List<string> rootFolders = new List<string>();
+            rootFolders.Add(Application.dataPath);
+            rootFolders.Add(Path.GetFullPath("Packages"));
+            searchForFoldersTask = Task.Run(() => SearchForFoldersAsync(rootFolders));
```
Comment (julenka) [inline Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:223]: In my testing, I found one more set of folders we need to search, the package cache. I think you need to add something like:

```
rootFolders.Add(Path.GetFullPath(Path.Combine("Library", "PackageCache")));
```

Comment (keveleigh) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:223]: @davidkline-ms I think this another data point for using the `AssetDatabase` APIs instead of the way the MixedRealityToolkitFiles class is currently written. Those APIs _should_ search across all Assets and Packages and we can then get the proper loading path from that.

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WMR.ConfigCheck.asmdef (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WMR.ConfigCheck.asmdef]: In fact, there were some unnecessary assemblies that have been folded in this PR. :smile:

File: Assets/MRTK/Core/AssemblyInfo.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/AssemblyInfo.cs]: Added a custom preserve attribute that supports inheritance. this allows us to decorate only the base class(es) to perform strategic, targeted preservation

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 3 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:223]: @keveleigh is working to update our MixedRealityToolkitFiles class to use AssetDatabase APIs. Once that work is complete, this PR will be modified to adapt to the change.

**Comment by julenka** [comment]
@davidkline-ms I still want to stress that when I've been doing my own testing using mrtk as a package, **using the [Preserve] attribute was not enough**. I needed to use link.xml in Assets folder to ensure that Unity copied over the MRTK dlls that contained WindowsMixedRealityDeviceManager. Without copying over these dlls I got errors like "Unable to register IMixedRealityInputDeviceManager service with a null concrete type."

**Comment by julenka** [comment]
@davidkline-ms and @keveleigh it would be very helpful for devs if we also added the following lines in MixedRealityInputSystem.cs:

```
                    if (configuration.ComponentType.Type == null)
                    {
                        Debug.LogError($"Component Type is null for {configuration.ComponentName}. Check if this component has been stripped from the build. See: https://docs.unity3d.com/2018.4/Documentation/Manual/ManagedCodeStripping.html");
                    }
```

See [the original UPM package PR](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4616/files). 

That change makes it much easier to identify which specific classes are missing from the compiled project.

**Comment by davidkline-ms** [comment]
> @davidkline-ms I still want to stress that when I've been doing my own testing using mrtk as a package, using the [Preserve] attribute was not enough. I needed to use link.xml in Assets folder to ensure that Unity copied over the MRTK dlls that contained WindowsMixedRealityDeviceManager. Without copying over these dlls I got errors like "Unable to register IMixedRealityInputDeviceManager service with a null concrete type."

Understood. The strange thing is that I am _not_ getting these errors in my testing. It looks like there may be something else going on here. Unfortunately the link.xml file results in slow build times as the entire assembly is left unstripped.

I will continue to investigate.

**Comment by davidkline-ms** [comment]
@julenka, maybe we need to add this to our assemblies...

```
[assembly: UnityEngine.Scripting.AlwaysLinkAssembly]
```

**Comment by davidkline-ms** [comment]
> @davidkline-ms and @keveleigh it would be very helpful for devs if we also added the following lines in MixedRealityInputSystem.cs:

@julenka, this is going to take a bit more modification as the BaseDataProviderAccessCoreSystemClass behaves quite differently than the 2.0 input system.

**Comment by davidkline-ms** [comment]
> @davidkline-ms and @keveleigh it would be very helpful for devs if we also added the following lines in MixedRealityInputSystem.cs:
@julenka, this is going to take a bit more modification as the BaseDataProviderAccessCoreSystemClass behaves quite differently than the 2.0 input system.

I just pushed a fix for this. Adds new protected overload and obsoletes the old one.

File: Assets/MRTK/Core/Attributes/PreserveAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        AttributeTargets.Enum |
+        AttributeTargets.Constructor |
+        AttributeTargets.Method |
+        AttributeTargets.Property |
+        AttributeTargets.Field |
+        AttributeTargets.Event |
+        AttributeTargets.Interface |
+        AttributeTargets.Delegate,
+        Inherited = true)]
+    public class MixedRealityToolkitPreserveAttribute : UnityEngine.Scripting.PreserveAttribute
```
Comment (wiwei) [inline Assets/MRTK/Core/Attributes/PreserveAttribute.cs:24]: I don't see this being used anywhere in this PR - I take it this is primarily for future use?

Can we add some more doccomments about when you should use this/what the purpose is etc?

File: Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs [hunk 5]
```diff
@@ -152,12 +152,31 @@ public virtual IMixedRealityDataProvider GetDataProvider(string name)
         /// </summary>
         protected bool RegisterDataProvider<T>(
             Type concreteType,
+            string providerName,
             SupportedPlatforms supportedPlatforms = (SupportedPlatforms)(-1),
             params object[] args) where T : IMixedRealityDataProvider
         {
             return RegisterDataProviderInternal<T>(
-                true, // Retry with an added IMixedRealityService parameter
+                true, // Register the provider
```
Comment (wiwei) [inline Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:160]: Why the comment change from retry -> register? The parameter name is retryWithRegistrar + the name of this function is register so I would assume that it's trying to register.

**Comment by wiwei** [comment]
Taking a look at MixedRealityToolkitPreserve, I still feel like long term this is the incorrect way to go down. This will make it so that any build will always preserve/not optimize away all of the services, even when other services are not used.

While this is what happens with link.xml today, with link.xml it's possible for a consumer to actually pick and choose what gets optimized away and what doesn't, and we can actually build tooling that will update a link.xml to work correctly given their current profiles.

With this change, it will be impossible for a consumer to change this behavior because if they go with:
UPM: source code is read only
Precompiled binaries: obviously they can't change the source code.

Even with .unitypackage they will still have to modify the code themselves and then get everything overwritten next time they upgrade.

Given my previous comment

> Note that it's possible to have a link.xml that references OTHER assemblies:
> https://docs.unity3d.com/Manual/ManagedCodeStripping.html
> Note that link.xml files are not supported inside packages, but you can reference package ssemblies from non-package link.xml files.

Unless that doesn't work, I really, really, really don't think we should do things this way because this will close the door on future optimizations.

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs [hunk 6]
```diff
@@ -138,8 +138,42 @@ public static void OnPostprocessAllAssets(string[] importedAssets, string[] dele
         /// <returns>The project relative path.</returns>
         /// <remarks>This doesn't produce paths that contain step out '..' relative paths.</remarks>
         public static string GetAssetDatabasePath(string absolutePath)
+        {
+            string assetDatabasePath = Path.GetFullPath(absolutePath).Replace("\\", "/");
```
Comment (wiwei) [inline Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:142]: Was the path normalization required to get things to work here?

Comment (wiwei) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:142]: Ah, to deal with the separator index in PackageCache

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs [hunk 7]
```diff
@@ -138,8 +138,42 @@ public static void OnPostprocessAllAssets(string[] importedAssets, string[] dele
         /// <returns>The project relative path.</returns>
         /// <remarks>This doesn't produce paths that contain step out '..' relative paths.</remarks>
         public static string GetAssetDatabasePath(string absolutePath)
```
Comment (wiwei) [inline Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:140]: Please add tests for this function.

I think that dealing with Path.GetFullPath is going to be tricky (aside from creating temp files in tests) but the other stuff would be important to have test coverage.

The string manipulation below is "simple" but easy to regress, and I can't imagine how scary it would be to make a change in the future and hope that we don't break one of the three clauses.

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs [hunk 8]
```diff
@@ -371,15 +410,21 @@ public static string GetGeneratedFolder
...
                 searchForFoldersToken.Cancel();
             }
 
             searchForFoldersToken = new CancellationTokenSource();
-            await Task.Run(() => SearchForFolders(rootPath, searchForFoldersToken.Token), searchForFoldersToken.Token);
+            await Task.Run(() =>
+                {
+                    for (int i = 0; i < rootFolders.Count; i++)
+                    {
+                        SearchForFolders(rootFolders[i], searchForFoldersToken.Token);
```
Comment (wiwei) [inline Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:425]: Should we break if we ended up finding it (i.e. rather than keep searching)

File: Assets/MRTK/Providers/UnityAR/Microsoft.MixedReality.Toolkit.Providers.UnityAR.asmdef [hunk 9]
```diff
@@ -2,8 +2,7 @@
     "name": "Microsoft.MixedReality.Toolkit.Providers.UnityAR",
     "references": [
         "Microsoft.MixedReality.Toolkit",
-        "Microsoft.MixedReality.Toolkit.Async",
-        "Unity.XR.ARFoundation"
```
Comment (wiwei) [inline Assets/MRTK/Providers/UnityAR/Microsoft.MixedReality.Toolkit.Providers.UnityAR.asmdef:6]: Is this change related to the UPM stuff?

**Comment by wiwei** [comment]
Looking at this change, I'm good with the file-finding code changes (modulo getting test coverage for it).

The linking/preserving one I think is the wrong path to go down, though I can be convinced otherwise if the majority of the team wants to go down that path and is able to come up with solutions for how to avoid the issues I highlighted.

File: Assets/MRTK/Providers/UnityAR/Microsoft.MixedReality.Toolkit.Providers.UnityAR.asmdef (hunk 9 above)
Comment (davidkline-ms) [reply Assets/MRTK/Providers/UnityAR/Microsoft.MixedReality.Toolkit.Providers.UnityAR.asmdef:6]: it is not. must have been accidental. thanks for the catch!

**Comment by davidkline-ms** [comment]
> Unless that doesn't work, I really, really, really don't think we should do things this way because this will close the door on future optimizations.

Fair enough. The goal was to reduce the code being preserved without putting the burden on the customer copying/creating an appropriate link.xml file.

This _should_ include far less code than the current link file.

File: Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs (hunk 5 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Services/BaseDataProviderAccessCoreSystem.cs:160]: mistake :)

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 8 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:425]: I'll take a look

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 7 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:140]: Added

**Comment by davidkline-ms** [comment]
Removed code preservation changes from this PR. I have intentionally left the AlwaysLinkAssembly attributes to ensure that Unity processes the MRTK services and data providers during linking.

#8092 has been created to enable flexible, customer configurable code preservation in a future PR.

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 8 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:425]: None of the methods in the call stack return or early exit as currently written. I will look at adding this.

Fortunately, the ones needing modification are all private. There will be no breaking changes involved.

Comment (keveleigh) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:425]: Oof yeah, none of the called methods provide any info on whether they were successful or not. Might be able to pipe a bool all the way back up, but I think some parts of this class need to be rewritten for robustness / performance anyway. Probably best to do it then.

Comment (davidkline-ms) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:425]: The scope of the required changes feel outside of that of this PR. There are some structural changes in MixedRealityToolkitFiles.cs that need investigation.

File: Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs [hunk 10]
```diff
@@ -8,3 +8,7 @@
 
 [assembly: AssemblyProduct("Microsoft® Mixed Reality Toolkit Extensions")]
 [assembly: AssemblyCopyright("Copyright © Microsoft Corporation")]
+
+[assembly: UnityEngine.Scripting.Preserve]
```
Comment (wiwei) [inline Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs:12]: This still has a preserve marked on it

Comment (wiwei) [reply Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs:12]: It might be worth doing another pass (i.e. grep -ri preserve *) to make sure we fully clean this.

File: Assets/MRTK/Tests/EditModeTests/Core/MixedRealityToolkitFilesTests.cs [hunk 11]
```diff
@@ -111,6 +112,31 @@ public void TestFileUtilitiesFindFilesInAssets()
...
+        public void VerifyGetAssetDatabasePath()
+        {
+            // File in the Assets folder structure
+            string inputPath = @"c:\projects\MyUnityProject\Assets\materials\test.mat";
+            string expectedPath = "Assets/materials/test.mat";
+            string actualPath = MixedRealityToolkitFiles.GetAssetDatabasePath(inputPath);
+            Asset.Equals(expectedPath, actualPath);
+
+            // File in the Packages folder structire
+            inputPath = "d:/source/projects/Packages/CoolFeature/textures/background.png";
```
Comment (wiwei) [inline Assets/MRTK/Tests/EditModeTests/Core/MixedRealityToolkitFilesTests.cs:128]: Is the / intentional here vs \ from the previous one?

Comment (wiwei) [reply Assets/MRTK/Tests/EditModeTests/Core/MixedRealityToolkitFilesTests.cs:128]: I'm guessing yes to test the other slash to verify behavior

**Comment by wiwei** [comment]
I see that link.xml is still deleted - this will pretty much immediately regress people who use precompiled drops right? If I'm reading the AlwaysLinkAssembly documentation right it doesn't look like it will do the same thing.

File: Assets/MRTK/Tests/EditModeTests/Core/MixedRealityToolkitFilesTests.cs (hunk 11 above)
Comment (davidkline-ms) [reply Assets/MRTK/Tests/EditModeTests/Core/MixedRealityToolkitFilesTests.cs:128]: correct. it is to verify the behavior regardless of slash direction.

**Comment by davidkline-ms** [comment]
>I see that link.xml is still deleted - this will pretty much immediately regress people who use precompiled drops right? If I'm reading the AlwaysLinkAssembly documentation right it doesn't look like it will do the same thing.

Good call. i will restore it.

File: Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs (hunk 10 above)
Comment (davidkline-ms) [reply Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs:12]: i missed one file 😄

Comment (davidkline-ms) [reply Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs]: fixed

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs [hunk 12]
```diff
@@ -138,8 +138,42 @@ public static void OnPostprocessAllAssets(string[] importedAssets, string[] dele
         /// <returns>The project relative path.</returns>
         /// <remarks>This doesn't produce paths that contain step out '..' relative paths.</remarks>
         public static string GetAssetDatabasePath(string absolutePath)
+        {
+            string assetDatabasePath = Path.GetFullPath(absolutePath).Replace("\\", "/");
+            string token = string.Empty;
+            string newRoot = string.Empty;
+            if (assetDatabasePath.Contains("Assets"))
```
Comment (julenka) [inline Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:145]: There's an edge case I found that this does not work for:

1. Texture you are looking for is under "StandardAssets" subfolder, under the "PackageCache" folder.

this set of if statements sets the token to be "assets" in taht case.

File: Assets/MRTK/Providers/LeapMotion/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.LeapMotion.ConfigurationChecker.asmdef [hunk 13]
```diff
@@ -1,16 +0,0 @@
-{
-    "name": "Microsoft.MixedReality.Toolkit.Providers.LeapMotion.ConfigurationChecker",
```
Comment (keveleigh) [inline Assets/MRTK/Providers/LeapMotion/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.LeapMotion.ConfigurationChecker.asmdef:2]: Is this removal combining the config checker and editor assemblies for Leap Motion?

File: Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs [hunk 14]
```diff
@@ -8,3 +8,5 @@
 
 [assembly: AssemblyProduct("Microsoft® Mixed Reality Toolkit Extensions")]
 [assembly: AssemblyCopyright("Copyright © Microsoft Corporation")]
+
+[assembly: UnityEngine.Scripting.AlwaysLinkAssembly]
```
Comment (keveleigh) [inline Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs:12]: Sorry if I missed a conversation on this entirely, but what do these do now? I thought we were reverting to the link.xml?

File: Assets/MRTK/Tests/EditModeTests/Core/MixedRealityToolkitFilesTests.cs [hunk 15]
```diff
@@ -111,6 +112,31 @@ public void TestFileUtilitiesFindFilesInAssets()
             Assert.IsTrue(files.Length == 1);
         }
 
+        /// <summary>
+        /// Validates that MixedRealutyToolkit.GetAssetDatabasePath produces the correct asset path.
```
Comment (keveleigh) [inline Assets/MRTK/Tests/EditModeTests/Core/MixedRealityToolkitFilesTests.cs:116]:
```suggestion
        /// Validates that MixedRealityToolkit.GetAssetDatabasePath produces the correct asset path.
```

File: Assets/link.xml [hunk 16]
```diff
@@ -1,27 +1,27 @@
...
-  <assembly fullname="Microsoft.MixedReality.Toolkit.SDK" preserve="all"/>
-  <assembly fullname="Microsoft.MixedReality.Toolkit.Services.BoundarySystem" preserve="all"/>
-  <assembly fullname="Microsoft.MixedReality.Toolkit.Services.CameraSystem" preserve="all"/>
-  <assembly fullname="Microsoft.MixedReality.Toolkit.Services.DiagnosticsSystem" preserve="all"/>
-  <assembly fullname="Microsoft.MixedReality.Toolkit.Services.InputSystem" preserve="all"/>
-  <assembly fullname="Microsoft.MixedReality.Toolkit.Services.SceneSystem" preserve="all"/>
-  <assembly fullname="Microsoft.MixedReality.Toolkit.Services.SpatialAwarenessSystem" preserve="all"/>
-  <assembly fullname="Microsoft.MixedReality.Toolkit.Services.TeleportSystem" preserve="all"/>
-</linker>
+<linker>
```
Comment (keveleigh) [inline Assets/link.xml:1]: Line endings change or something?

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WindowsMixedReality.ConfigurationChecker.asmdef [hunk 17]
```diff
@@ -1,16 +0,0 @@
-{
-    "name": "Microsoft.MixedReality.Toolkit.Providers.WindowsMixedReality.ConfigCheck",
-    "references": [
-        "Microsoft.MixedReality.Toolkit.Editor.Utilities"
```
Comment (keveleigh) [inline Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WindowsMixedReality.ConfigurationChecker.asmdef:4]: Not this assembly, but the others that have their file names change (couldn't figure out how to comment on a renamed but not changed file):

I believe Unity serializes these references by file name, not by actual assembly name. It'd be good to make sure none of the renamed Editor assemblies are referenced anywhere else.

File: Assets/MRTK/Providers/LeapMotion/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.LeapMotion.ConfigurationChecker.asmdef (hunk 13 above)
Comment (keveleigh) [reply Assets/MRTK/Providers/LeapMotion/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.LeapMotion.ConfigurationChecker.asmdef:2]: Ah I see that the config checkers were all moved to the editor assembly. Seems fine to me! I think they were originally separate to try to help these scripts load even if a different assembly was broken, but it seems like that doesn't happen anyway (and editor assemblies _should_ be mostly package agnostic at the moment).

File: Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WindowsMixedReality.ConfigurationChecker.asmdef (hunk 17 above)
Comment (keveleigh) [reply Assets/MRTK/Providers/WindowsMixedReality/Shared/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.WindowsMixedReality.ConfigurationChecker.asmdef:4]: Actually I think I'm wrong here. I thought I had run into issues when file names changed, but I'm not seeing that at the moment.

File: Assets/MRTK/Providers/LeapMotion/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.LeapMotion.ConfigurationChecker.asmdef (hunk 13 above)
Comment (wiwei) [reply Assets/MRTK/Providers/LeapMotion/Editor/ConfigurationChecker/Microsoft.MixedReality.Toolkit.Providers.LeapMotion.ConfigurationChecker.asmdef:2]: @CDiaz-MS FYI

File: Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs (hunk 14 above)
Comment (davidkline-ms) [reply Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs:12]: This ensures the assembly is always processed by the linker. With the forthcoming PR to add an mrtk specific linker to generated folder, i can likely remove these.

File: Assets/link.xml (hunk 16 above)
Comment (davidkline-ms) [reply Assets/link.xml:1]: i had deleted it in a previous commit, then re-added from the mrtk_development branch. Not sure what GH is flagging here.

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 12 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs:145]: Good catch! Thanks @julenka! I'll add a fix (and test case)

File: Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs (hunk 14 above)
Comment (davidkline-ms) [reply Assets/MRTK/Extensions/SceneTransitionService/AssemblyInfo.cs]: removed

File: Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 12 above)
Comment (davidkline-ms) [reply Assets/MRTK/Core/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs]: fixed and new edge case tests added
